src/number_theory/lucas_lehmer.lean

```diff
@@ -0,0 +1,558 @@
+  Lucas_Lehmer_residue p = 0 ↔ s_mod p (p-2) = 0 :=
+begin
+  dsimp [Lucas_Lehmer_residue],
+  rw s_zmod_eq_s_mod p,
+  split,
+  { -- We want to use that fact that `0 ≤ s_mod p (p-2) < 2^p - 1`
+    -- and `Lucas_Lehmer_residue p = 0 → 2^p - 1 ∣ s_mod p (p-2)`.
+    intro h,
+    simp [zmod.int_coe_zmod_eq_zero_iff_dvd] at h,
+    apply int.eq_zero_of_dvd_of_nonneg_of_lt _ _ h; clear h,
+    apply s_mod_nonneg _ (nat.lt_of_succ_lt w),
+    convert s_mod_lt _ (nat.lt_of_succ_lt w) (p-2),
+    push_cast [nat.one_le_two_pow p],
+    refl, },
+  { intro h, rw h, simp, },
+end
+
+/--
+A Mersenne number 2^p-1 is prime if and only if
+the Lucas-Lehmer residue `s p (p-2) % (2^p - 1)` is zero.
+-/
+@[derive decidable_pred]
+def Lucas_Lehmer_test (p : ℕ) := Lucas_Lehmer_residue p = 0
+
+/-- `q` is defined as the minimum factor of `Mersenne p`, bundled as an `ℕ+`. -/
+def q (p : ℕ) : ℕ+ := ⟨nat.min_fac (Mersenne p), nat.min_fac_pos (Mersenne p)⟩
+
+instance fact_pnat_pos (q : ℕ+) : fact (0 < (q : ℕ)) :=
+q.2
+
+/-- We construct the ring `X q` as ℤ/qℤ + √3 ℤ/qℤ. -/
+@[derive [add_comm_group, decidable_eq, fintype]]
+def X (q : ℕ+) := (zmod q) × (zmod q)
+
+namespace X
+variable {q : ℕ+}
+
+@[ext]
+lemma ext {x y : X q} (h₁ : x.1 = y.1) (h₂ : x.2 = y.2) : x = y :=
+begin
+  cases x, cases y,
+  congr; assumption
+end
+
+@[simp] lemma add_fst (x y : X q) : (x + y).1 = x.1 + y.1 := rfl
+@[simp] lemma add_snd (x y : X q) : (x + y).2 = x.2 + y.2 := rfl
+
+@[simp] lemma neg_fst (x : X q) : (-x).1 = -x.1 := rfl
+@[simp] lemma neg_snd (x : X q) : (-x).2 = -x.2 := rfl
+
+instance : has_mul (X q) :=
+{ mul := λ x y, (x.1*y.1 + 3*x.2*y.2, x.1*y.2 + x.2*y.1) }
+
+@[simp] lemma mul_fst (x y : X q) : (x * y).1 = x.1 * y.1 + 3 * x.2 * y.2 := rfl
+@[simp] lemma mul_snd (x y : X q) : (x * y).2 = x.1 * y.2 + x.2 * y.1 := rfl
+
+instance : has_one (X q) :=
+{ one := ⟨1,0⟩ }
+
+@[simp] lemma one_fst : (1 : X q).1 = 1 := rfl
+@[simp] lemma one_snd : (1 : X q).2 = 0 := rfl
+
+@[simp] lemma bit0_fst (x : X q) : (bit0 x).1 = bit0 x.1 := rfl
+@[simp] lemma bit0_snd (x : X q) : (bit0 x).2 = bit0 x.2 := rfl
+@[simp] lemma bit1_fst (x : X q) : (bit1 x).1 = bit1 x.1 := rfl
+@[simp] lemma bit1_snd (x : X q) : (bit1 x).2 = bit0 x.2 := by { dsimp [bit1], simp, }
+
+instance : monoid (X q) :=
+{ mul_assoc := λ x y z, by { ext; { dsimp, ring }, },
+  one := ⟨1,0⟩,
+  one_mul := λ x, by { ext; simp, },
+  mul_one := λ x, by { ext; simp, },
+  ..(infer_instance : has_mul (X q)) }
+
+lemma left_distrib (x y z : X q) : x * (y + z) = x * y + x * z :=
+by { ext; { dsimp, ring }, }
+
+lemma right_distrib (x y z : X q) : (x + y) * z = x * z + y * z :=
+by { ext; { dsimp, ring }, }
+
+instance : ring (X q) :=
+{ left_distrib := left_distrib,
+  right_distrib := right_distrib,
+  ..(infer_instance : add_comm_group (X q)),
+  ..(infer_instance : monoid (X q)) }
+
+instance : comm_ring (X q) :=
+{ mul_comm := λ x y, by { ext; { dsimp, ring }, },
+  ..(infer_instance : ring (X q))}
```

Comment:
Is there a reason why you don't just merge all of these into one? Even the proofs of `left/right_distrib` are one-liners...

Reply:
1. it was better pedagogically for the student to do it step by step, and I never minimised
2. when I today tried to minimise it, this [lean issue](https://github.com/leanprover-community/lean/issues/197) caused trouble, with expressions like `(add_comm_group.add y z)` appearing instead of `y + z`.

(It's possible, although I haven't heard either way, that [a new PR](https://github.com/leanprover-community/lean/pull/282) is about to fix this issue.)